### ISSUE TITLE
JDK11 reflection warnings

### ISSUE THREAD

**Issue by DagW** [jolokia/jolokia#394]
We recently updated our application to run with AdoptOpenJDK 11,
upon start the following lines appeared

```
Oct 17 11:29:14 XXX-01 java[58654]: WARNING: An illegal reflective access operation has occurred
Oct 17 11:29:14 XXX-01 java[58654]: WARNING: Illegal reflective access by org.jolokia.jvmagent.client.command.AbstractBaseCommand (file:/usr/local/libexec/jolokia-jvm-agent.jar) to method sun.tools.attach.HotSpotVirtualMachine.getSystemProperties()
Oct 17 11:29:14 XXX-01 java[58654]: WARNING: Please consider reporting this to the maintainers of org.jolokia.jvmagent.client.command.AbstractBaseCommand
Oct 17 11:29:14 XXX-01 java[58654]: WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
Oct 17 11:29:14 XXX-01 java[58654]: WARNING: All illegal access operations will be denied in a future release
```

Thoughts?

**Comment by bfg111** [comment]
Seeing this error as well
Java version:
```
java 11.0.1 2018-10-16 LTS
Java(TM) SE Runtime Environment 18.9 (build 11.0.1+13-LTS)
Java HotSpot(TM) 64-Bit Server VM 18.9 (build 11.0.1+13-LTS, mixed mode)
```

Jolokia version:
```
Jolokia JVM Agent 1.6.0
```

```
Mar 22 14:54:50 xxxxxxxx java[26349]: WARNING: An illegal reflective access operation has occurred
Mar 22 14:54:50 xxxxxxxx java[26349]: WARNING: Illegal reflective access by org.jolokia.jvmagent.client.command.AbstractBaseCommand (file:/[base_location]/jolokia-jvm-1.6.0-agent.jar) to method sun.tools.attach.HotSpotVirtualMachine.getSystemProperties()
Mar 22 14:54:50 xxxxxxxx java[26349]: WARNING: Please consider reporting this to the maintainers of org.jolokia.jvmagent.client.command.AbstractBaseCommand
Mar 22 14:54:50 xxxxxxxx java[26349]: WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
Mar 22 14:54:50 xxxxxxxx java[26349]: WARNING: All illegal access operations will be denied in a future release
Mar 22 14:54:51 xxxxxxxx java[26349]: InvocationTargetException (command: start)
Mar 22 14:54:51 xxxxxxxx java[26349]: java.lang.reflect.InvocationTargetException
Mar 22 14:54:51 xxxxxxxx java[26349]: at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
Mar 22 14:54:51 xxxxxxxx java[26349]: at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
Mar 22 14:54:51 xxxxxxxx java[26349]: at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
Mar 22 14:54:51 xxxxxxxx java[26349]: at java.base/java.lang.reflect.Method.invoke(Method.java:566)
Mar 22 14:54:51 xxxxxxxx java[26349]: at org.jolokia.jvmagent.client.command.AbstractBaseCommand.loadAgent(AbstractBaseCommand.java:72)
Mar 22 14:54:51 xxxxxxxx java[26349]: at org.jolokia.jvmagent.client.command.StartCommand.execute(StartCommand.java:48)
Mar 22 14:54:51 xxxxxxxx java[26349]: at org.jolokia.jvmagent.client.command.CommandDispatcher.dispatchCommand(CommandDispatcher.java:78)
Mar 22 14:54:51 xxxxxxxx java[26349]: at org.jolokia.jvmagent.client.AgentLauncher.main(AgentLauncher.java:62)
Mar 22 14:54:51 xxxxxxxx java[26349]: Caused by: com.sun.tools.attach.AgentInitializationException: Agent JAR loaded but agent failed to initialize
Mar 22 14:54:51 xxxxxxxx java[26349]: at jdk.attach/sun.tools.attach.HotSpotVirtualMachine.loadAgent(HotSpotVirtualMachine.java:165)
Mar 22 14:54:51 xxxxxxxx java[26349]: ... 8 more
```

**Comment by rhuss** [comment]
Tbh, Jolokia has not yet been ported to Java 11 (nor really tested). This is on the todo list, but currently, I'm buried under work so that this has no high priority for me at the moment.

However, if anyone feels fancy and tries to port Jolokia to Java 11 (including test) I would be more than happy in integrating pull requests.

**Comment by stevedwray** [comment]
Is there any update on java 11 support in jolokia?

**Comment by grgrzybek** [comment]
I'm not sure if it's the right analysis, but, here it is.

In first attempt I started simple application using:
```
$ java -javaagent:/home/ggrzybek/.m2/repository/org/jolokia/jolokia-jvm/1.6.0/jolokia-jvm-1.6.0-agent.jar -cp target/jmx-application.jar grgr.javase.jdk11.JmxApplication
Press <ENTER> to quit
I> No access restrictor found, access to any MBean is allowed
Jolokia: Agent started with URL http://127.0.0.1:8778/jolokia/
```

And it just worked:
```
$ curl -s http://localhost:8778/jolokia/read/JMImplementation:type=MBeanServerDelegate/ImplementationVersion|jq .
{
  "request": {
    "mbean": "JMImplementation:type=MBeanServerDelegate",
    "attribute": "ImplementationVersion",
    "type": "read"
  },
  "value": "11.0.1+13-LTS",
  "timestamp": 1608561124,
  "status": 200
}
```

But after checking the stack trace, I found it's a bit different scenario, so I've started the same application but without `-javaagent`:
```
$ java -cp target/jmx-application.jar grgr.javase.jdk11.JmxApplication
Press <ENTER> to quit
```

And reproduced the problem with:
```
$ java -cp /home/ggrzybek/.m2/repository/org/jolokia/jolokia-jvm/1.6.0/jolokia-jvm-1.6.0-agent.jar org.jolokia.jvmagent.client.AgentLauncher start 228927
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.jolokia.jvmagent.client.command.AbstractBaseCommand (file:/data/ggrzybek/.m2/repository/org/jolokia/jolokia-jvm/1.6.0/jolokia-jvm-1.6.0-agent.jar) to method sun.tools.attach.HotSpotVirtualMachine.getSystemProperties()
WARNING: Please consider reporting this to the maintainers of org.jolokia.jvmagent.client.command.AbstractBaseCommand
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
Started Jolokia for PID 228927
http://127.0.0.1:8778/jolokia/
```

However, the agent still started in the `228927` application:
```
$ curl -s http://localhost:8778/jolokia/read/JMImplementation:type=MBeanServerDelegate/ImplementationVersion|jq .
{
  "request": {
    "mbean": "JMImplementation:type=MBeanServerDelegate",
    "attribute": "ImplementationVersion",
    "type": "read"
  },
  "value": "11.0.1+13-LTS",
  "timestamp": 1608561344,
  "status": 200
}
```

However,  because jolokia is _not_ a JDK9+ module, it should be started from classpath anyway, thus as a part of _unnamed_ module. So we can tweak the module setup a bit:
```
$ java --add-opens=jdk.attach/sun.tools.attach=ALL-UNNAMED -cp /home/ggrzybek/.m2/repository/org/jolokia/jolokia-jvm/1.6.0/jolokia-jvm-1.6.0-agent.jar org.jolokia.jvmagent.client.AgentLauncher start 21193
Started Jolokia for PID 21193
http://127.0.0.1:8778/jolokia/
```

All that's needed is:
```
--add-opens=jdk.attach/sun.tools.attach=ALL-UNNAMED
```

So Jolokia can reflectively access classes from `sun.tools.attach` package of `jdk.attach` module.

**Comment by rhuss** [comment]
@grgrzybek thanks for the investigation! I agree that this is the cause. The question is, whether we should remove the reflective access to JDK.attach as this was only used to check dynamically whether this package is available (fair). So import that classes would help here, with the disadvantage that the runtime error is probably a bit more ugly.

wdyt ?

**Comment by grgrzybek** [comment]
With:
```
<plugin>
  <artifactId>maven-surefire-plugin</artifactId>
  <configuration>
    <argLine>--illegal-access=warn</argLine>
  </configuration>
</plugin>
```

I got all the reflection problems when running all the tests under JDK 11:
```
WARNING: Illegal reflective access by org.easymock.cglib.core.ReflectUtils$1 to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Illegal reflective access by org.jolokia.jmx.JolokiaMBeanServer to field javax.management.modelmbean.RequiredModelMBean.managedResource
WARNING: Illegal reflective access by org.jolokia.jvmagent.client.util.VirtualMachineHandler to method sun.tools.attach.VirtualMachineImpl.detach()
WARNING: Illegal reflective access by org.jolokia.jvmagent.security.KeyStoreUtil to method sun.security.tools.keytool.CertAndKeyGen.getSelfCertificate(sun.security.x509.X500Name,java.util.Date,long)
WARNING: Illegal reflective access by org.jolokia.util.ClassUtil to constructor sun.security.tools.keytool.CertAndKeyGen(java.lang.String,java.lang.String)
WARNING: Illegal reflective access by org.jolokia.util.ClassUtil to constructor sun.security.x509.X500Name(java.lang.String,java.lang.String,java.lang.String,java.lang.String,java.lang.String,java.lang.String)
WARNING: Illegal reflective access by org.jolokia.util.ClassUtil to method sun.security.tools.keytool.CertAndKeyGen.generate(int)
WARNING: Illegal reflective access by org.jolokia.util.ClassUtil to method sun.security.tools.keytool.CertAndKeyGen.getPrivateKey()
```

I got rid of all the warnings with (again in maven-surefire-plugin configuration) with:
```
--add-opens java.base/java.lang=ALL-UNNAMED
--add-opens java.management/javax.management.modelmbean=ALL-UNNAMED
--add-opens jdk.attach/sun.tools.attach=ALL-UNNAMED
--add-opens java.base/sun.security.tools.keytool=ALL-UNNAMED
--add-opens java.base/sun.security.x509=ALL-UNNAMED
```

I'll check how many of these reflection accesses can be replaced by normal invocation (thinking about Quarkus as well) - but that's next year!

**Comment by rhuss** [comment]
@grgrzybek great findings and thanks for checking it out ! 

'wish you an awesome recharge time and see you next year !

**Comment by grgrzybek** [comment]
@rhuss a bit of progress. I think I can handle all the cases that are causing warnings on JDK11 (when not using `--add-opens`), **except one**.
The problem is `org.jolokia.jmx.JolokiaMBeanServer#extractJsonMBeanAnnotation()` which checks if the mbean is `javax.management.modelmbean.ModelMBean`.
In Jolokia itself, only one of the tests (`org.jolokia.jmx.JolokiaMBeanServerTest#withModelMBean()`) registers a `javax.management.modelmbean.RequiredModelMBean` wrapping an object annotated with `org.jolokia.jmx.JsonMBean` and expect it to work, so it HAS to reflectively access the bean.

There are three solutions:
* `--add-opens java.management/javax.management.modelmbean=ALL-UNNAMED`
* convert `@JsonMBean` values into `javax.management.ObjectName` key/value pairs (potentially making a copy of the original `ObjectName`)
* do not bother getting `@JsonMBean` in `org.jolokia.jmx.JolokiaMBeanServer#registerMBean()` if the object is already a `RequiredModelMBean`

I've already altered `org.jolokia.jvmagent.client.util.VirtualMachineHandler` using Multi-Release Jar and JDK9+ specific version of this class (to be polished, but it already works on Oracle JDK, OpenJDK and IBM JDK).

easymock + cglib may be a problem on JDK17 which already hidden `sun.misc.Unsafe`.

key/cert Jolokia tools using reflection can be rewritten in non-reflective way (work in progress).

**Comment by grgrzybek** [comment]
Handling non-reflective access to `com.sun.tools.attach.VirtualMachine`...

Normally, on JDK11 with jolokia-jvm-1.6.2.jar, when attaching to a JVM process we can see all WARNed access places using:
```
 java --illegal-access=debug -cp jolokia-jvm-1.6.2.jar org.jolokia.jvmagent.client.AgentLauncher start 115163
WARNING: Illegal reflective access by org.jolokia.jvmagent.client.util.VirtualMachineHandler (file:jolokia-jvm-1.6.2.jar) to method sun.tools.attach.HotSpotVirtualMachine.getSystemProperties()
	at org.jolokia.jvmagent.client.util.VirtualMachineHandler.getSystemProperties(VirtualMachineHandler.java:171)
	at org.jolokia.jvmagent.client.command.AbstractBaseCommand.getAgentSystemProperties(AbstractBaseCommand.java:114)
	at org.jolokia.jvmagent.client.command.AbstractBaseCommand.checkAgentUrl(AbstractBaseCommand.java:103)
	at org.jolokia.jvmagent.client.command.AbstractBaseCommand.checkAgentUrl(AbstractBaseCommand.java:82)
	at org.jolokia.jvmagent.client.command.StartCommand.execute(StartCommand.java:42)
	at org.jolokia.jvmagent.client.command.CommandDispatcher.dispatchCommand(CommandDispatcher.java:78)
	at org.jolokia.jvmagent.client.AgentLauncher.main(AgentLauncher.java:62)
WARNING: Illegal reflective access by org.jolokia.jvmagent.client.util.VirtualMachineHandler (file:jolokia-jvm-1.6.2.jar) to method sun.tools.attach.HotSpotVirtualMachine.loadAgent(java.lang.String,java.lang.String)
	at org.jolokia.jvmagent.client.util.VirtualMachineHandler.loadAgent(VirtualMachineHandler.java:156)
	at org.jolokia.jvmagent.client.command.AbstractBaseCommand.loadAgent(AbstractBaseCommand.java:69)
	at org.jolokia.jvmagent.client.command.StartCommand.execute(StartCommand.java:45)
	at org.jolokia.jvmagent.client.command.CommandDispatcher.dispatchCommand(CommandDispatcher.java:78)
	at org.jolokia.jvmagent.client.AgentLauncher.main(AgentLauncher.java:62)
WARNING: Illegal reflective access by org.jolokia.jvmagent.client.util.VirtualMachineHandler (file:jolokia-jvm-1.6.2.jar) to method sun.tools.attach.HotSpotVirtualMachine.getSystemProperties()
	at org.jolokia.jvmagent.client.util.VirtualMachineHandler.getSystemProperties(VirtualMachineHandler.java:171)
	at org.jolokia.jvmagent.client.command.AbstractBaseCommand.getAgentSystemProperties(AbstractBaseCommand.java:114)
	at org.jolokia.jvmagent.client.command.AbstractBaseCommand.checkAgentUrl(AbstractBaseCommand.java:103)
	at org.jolokia.jvmagent.client.command.AbstractBaseCommand.checkAgentUrl(AbstractBaseCommand.java:82)
	at org.jolokia.jvmagent.client.command.StartCommand.execute(StartCommand.java:46)
	at org.jolokia.jvmagent.client.command.CommandDispatcher.dispatchCommand(CommandDispatcher.java:78)
	at org.jolokia.jvmagent.client.AgentLauncher.main(AgentLauncher.java:62)
WARNING: Illegal reflective access by org.jolokia.jvmagent.client.util.VirtualMachineHandler (file:jolokia-jvm-1.6.2.jar) to method sun.tools.attach.HotSpotVirtualMachine.getSystemProperties()
	at org.jolokia.jvmagent.client.util.VirtualMachineHandler.getSystemProperties(VirtualMachineHandler.java:171)
	at org.jolokia.jvmagent.client.command.AbstractBaseCommand.getAgentSystemProperties(AbstractBaseCommand.java:114)
	at org.jolokia.jvmagent.client.command.AbstractBaseCommand.checkAgentUrl(AbstractBaseCommand.java:103)
	at org.jolokia.jvmagent.client.command.StartCommand.execute(StartCommand.java:49)
	at org.jolokia.jvmagent.client.command.CommandDispatcher.dispatchCommand(CommandDispatcher.java:78)
	at org.jolokia.jvmagent.client.AgentLauncher.main(AgentLauncher.java:62)
Started Jolokia for PID 115163
http://127.0.0.1:8778/jolokia/
WARNING: Illegal reflective access by org.jolokia.jvmagent.client.util.VirtualMachineHandler (file:jolokia-jvm-1.6.2.jar) to method sun.tools.attach.VirtualMachineImpl.detach()
	at org.jolokia.jvmagent.client.util.VirtualMachineHandler.detachAgent(VirtualMachineHandler.java:97)
	at org.jolokia.jvmagent.client.AgentLauncher.main(AgentLauncher.java:71)
```

My incoming change detects Java version and leaves existing behavior when on JDK 8 or earlier (down to JDK 6), where `com.sun.tools.attach.VirtualMachine` is available through `tools.jar`).

However on JDK 9+ we have two scenarios:
1. Running on default JDK/JVM, where `jdk.attach` JPMS module is normally available
2. Running special JDK/JVM built using `jlink` and selected modules.

For scenario `#1` we have simply:
```
$ java --illegal-access=debug -cp jolokia-jvm-1.6.3-SNAPSHOT.jar org.jolokia.jvmagent.client.AgentLauncher start 115163
Started Jolokia for PID 115163
http://127.0.0.1:8778/jolokia/
```

For scenario `#2` we can prepare special, trimmed down JDK/JVM using:
```
$ jlink --add-modules=java.se --output /data/tmp/small-jdk

$ /data/tmp/small-jdk/bin/java --list-modules
java.base@11.0.11
java.compiler@11.0.11
java.datatransfer@11.0.11
java.desktop@11.0.11
java.instrument@11.0.11
java.logging@11.0.11
java.management@11.0.11
java.management.rmi@11.0.11
java.naming@11.0.11
java.net.http@11.0.11
java.prefs@11.0.11
java.rmi@11.0.11
java.scripting@11.0.11
java.se@11.0.11
java.security.jgss@11.0.11
java.security.sasl@11.0.11
java.sql@11.0.11
java.sql.rowset@11.0.11
java.transaction.xa@11.0.11
java.xml@11.0.11
java.xml.crypto@11.0.11
```

See? Only basic modules are available. On normal JDK there are:
```
$ java --list-modules
java.base@11.0.11
java.compiler@11.0.11
java.datatransfer@11.0.11
java.desktop@11.0.11
java.instrument@11.0.11
java.logging@11.0.11
java.management@11.0.11
java.management.rmi@11.0.11
java.naming@11.0.11
java.net.http@11.0.11
java.prefs@11.0.11
java.rmi@11.0.11
java.scripting@11.0.11
java.se@11.0.11
java.security.jgss@11.0.11
java.security.sasl@11.0.11
java.smartcardio@11.0.11
java.sql@11.0.11
java.sql.rowset@11.0.11
java.transaction.xa@11.0.11
java.xml@11.0.11
java.xml.crypto@11.0.11
jdk.accessibility@11.0.11
jdk.aot@11.0.11
jdk.attach@11.0.11
jdk.charsets@11.0.11
jdk.compiler@11.0.11
jdk.crypto.cryptoki@11.0.11
jdk.crypto.ec@11.0.11
jdk.dynalink@11.0.11
jdk.editpad@11.0.11
jdk.hotspot.agent@11.0.11
jdk.httpserver@11.0.11
jdk.internal.ed@11.0.11
jdk.internal.jvmstat@11.0.11
jdk.internal.le@11.0.11
jdk.internal.opt@11.0.11
jdk.internal.vm.ci@11.0.11
jdk.internal.vm.compiler@11.0.11
jdk.internal.vm.compiler.management@11.0.11
jdk.jartool@11.0.11
jdk.javadoc@11.0.11
jdk.jcmd@11.0.11
jdk.jconsole@11.0.11
jdk.jdeps@11.0.11
jdk.jdi@11.0.11
jdk.jdwp.agent@11.0.11
jdk.jfr@11.0.11
jdk.jlink@11.0.11
jdk.jshell@11.0.11
jdk.jsobject@11.0.11
jdk.jstatd@11.0.11
jdk.localedata@11.0.11
jdk.management@11.0.11
jdk.management.agent@11.0.11
jdk.management.jfr@11.0.11
jdk.naming.dns@11.0.11
jdk.naming.ldap@11.0.11
jdk.naming.rmi@11.0.11
jdk.net@11.0.11
jdk.pack@11.0.11
jdk.rmic@11.0.11
jdk.scripting.nashorn@11.0.11
jdk.scripting.nashorn.shell@11.0.11
jdk.sctp@11.0.11
jdk.security.auth@11.0.11
jdk.security.jgss@11.0.11
jdk.unsupported@11.0.11
jdk.unsupported.desktop@11.0.11
jdk.xml.dom@11.0.11
jdk.zipfs@11.0.11
```

So on this smaller JDK, we can see:
```
$ /data/tmp/small-jdk/bin/java --illegal-access=debug -cp jolokia-jvm-1.6.3-SNAPSHOT.jar org.jolokia.jvmagent.client.AgentLauncher start 115163
Can't load com.sun.tools.attach.VirtualMachine class. Is "jdk.attach" standard module available? (command: start) : com/sun/tools/attach/AgentLoadException
```

**Comment by grgrzybek** [comment]
#464 fixes one of the few places where reflection should be avoided in JDK9+.

The remaining warnings are:
```
WARNING: Illegal reflective access by org.easymock.cglib.core.ReflectUtils$1 (file:/home/ggrzybek/.m2/repository/org/easymock/easymock/4.3/easymock-4.3.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Illegal reflective access by org.jolokia.jmx.JolokiaMBeanServer (file:/data/sources/github.com/rhuss/jolokia/agent/jmx/target/classes/) to field javax.management.modelmbean.RequiredModelMBean.managedResource
WARNING: Illegal reflective access by org.jolokia.jvmagent.security.KeyStoreUtil (file:/data/sources/github.com/rhuss/jolokia/agent/jvm/target/classes/) to method sun.security.tools.keytool.CertAndKeyGen.getSelfCertificate(sun.security.x509.X500Name,java.util.Date,long)
WARNING: Illegal reflective access by org.jolokia.util.ClassUtil (file:/data/sources/github.com/rhuss/jolokia/agent/core/target/jolokia-core-1.6.3-SNAPSHOT.jar) to constructor sun.security.tools.keytool.CertAndKeyGen(java.lang.String,java.lang.String)
WARNING: Illegal reflective access by org.jolokia.util.ClassUtil (file:/data/sources/github.com/rhuss/jolokia/agent/core/target/jolokia-core-1.6.3-SNAPSHOT.jar) to constructor sun.security.x509.X500Name(java.lang.String,java.lang.String,java.lang.String,java.lang.String,java.lang.String,java.lang.String)
WARNING: Illegal reflective access by org.jolokia.util.ClassUtil (file:/data/sources/github.com/rhuss/jolokia/agent/core/target/jolokia-core-1.6.3-SNAPSHOT.jar) to method sun.security.tools.keytool.CertAndKeyGen.generate(int)
WARNING: Illegal reflective access by org.jolokia.util.ClassUtil (file:/data/sources/github.com/rhuss/jolokia/agent/core/target/jolokia-core-1.6.3-SNAPSHOT.jar) to method sun.security.tools.keytool.CertAndKeyGen.getPrivateKey()
```

**Comment by grgrzybek** [comment]
All tests pass on JDK17 as well after minor version changes to maven plugins.

**Comment by grgrzybek** [comment]
I've reimplemented self-signed certificate generation without using reflection API and `sun.security.tools.keytool.CertAndKeyGen`.
I've added basic ASN.1 generation (not parsing) to generate X.509 self-signed certificate according to RFC 5280.
All tests pass and now the remaining reflection warnings are:
```
17:19 $ find -name '*.dumpstream' | xargs cat | grep -v '^$' | grep -v '^# Created at' | grep '^WARNING:' | sort | uniq
WARNING: Illegal reflective access by org.easymock.cglib.core.ReflectUtils$1 (file:/home/ggrzybek/.m2/repository/org/easymock/easymock/4.3/easymock-4.3.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Illegal reflective access by org.jolokia.jmx.JolokiaMBeanServer (file:/data/sources/github.com/rhuss/jolokia/agent/jmx/target/classes/) to field javax.management.modelmbean.RequiredModelMBean.managedResource
```

**Comment by grgrzybek** [comment]
OK @rhuss :
* https://github.com/rhuss/jolokia/pull/464 is fixing reflective access to VM Attach API and (in lesser extent) is tweaking the reflective access to `RequiredModelMBean`
* https://github.com/rhuss/jolokia/pull/466 is adding non-reflective method to generate self-signed certificate without using any external library - all ASN.1 code is written from scratch based only on X.680/X.690 ITU standards

Tests pass on JDK8, JDK11 (without any illegal-access warnings) and JDK17 (after maven-compiler-plugin is switched from 1.6 to 1.7 - so it's rather JDK17 _ready_).

Easymock + Cglib require `--add-opens java.base/java.lang=ALL-UNNAMED` and it's only a test-scope problem. Should be fixed by switching to Mockito (I guess).

`org.jolokia.jmx.JolokiaMBeanServer#extractJsonMBeanAnnotation()` requires `--add-opens java.management/javax.management.modelmbean=ALL-UNNAMED` but only when user explicitly registers a `javax.management.modelmbean.RequiredModelMBean` MBean. This should be left as-is, because user will get proper WARNING message.

**Comment by rhuss** [comment]
Thanks, very cool ! Let me check the PR and get it ready asap.

**Comment by cniweb** [comment]
Hi @rhuss I hope you can check out the PR soon. The issue also affects some of our developers who use the OpenJDK image from the RedHat Container Catalog. Thanks!

**Comment by rhuss** [comment]
Sorry, I think we can close the issue as the changes has been already merged and released with Jolokia 1.7.1

@cniweb could you give 1.7.1 a test drive and let us know whether this works for you ? If so, we can close this issue.
